node/core/approval-voting/src/approval_checking.rs

```diff
@@ -0,0 +1,619 @@
+			n_approved + no_shows >= n_assigned
+		}
+	}
+}
+
+/// Determine the amount of tranches of assignments needed to determine approval of a candidate.
+pub fn tranches_to_approve(
+	approval_entry: &ApprovalEntry,
+	approvals: &BitSlice<BitOrderLsb0, u8>,
+	tranche_now: DelayTranche,
+	block_tick: Tick,
+	no_show_duration: Tick,
+	needed_approvals: usize,
+) -> RequiredTranches {
+	// This function progresses through a series of states while looping over the tranches
+	// that we are aware of. First, we perform an initial count of the number of assignments
+	// until we reach the number of needed assignments for approval. As we progress, we count the
+	// number of no-shows in each tranche.
+	//
+	// Then, if there are any no-shows, we proceed into a series of subsequent states for covering
+	// no-shows.
+	//
+	// We cover each no-show by a non-empty tranche, keeping track of the amount of further
+	// no-shows encountered along the way. Once all of the no-shows we were previously aware
+	// of are covered, we then progress to cover the no-shows we encountered while covering those,
+	// and so on.
+	#[derive(Debug)]
+	enum State {
+		// (assignments, no-shows)
+		InitialCount(usize, usize),
+		// (assignments, covered no-shows, covering no-shows, uncovered no-shows),
+		CoverNoShows(usize, usize, usize, usize),
+	}
+
+	impl State {
+		fn output(
+			&self,
+			tranche: DelayTranche,
+			tranche_now: DelayTranche,
+			needed_approvals: usize,
+			n_validators: usize,
+		) -> RequiredTranches {
+			match *self {
+				State::InitialCount(assignments, no_shows) =>
+					if assignments >= needed_approvals && no_shows == 0 {
+						RequiredTranches::Exact(tranche, 0)
+					} else {
+						// This happens only if there are not enough assignments, period.
+						//
+						// However, within the context of this function in particular,
+						// it can happen if there are enough assignments but there are no-shows.
+						// The calling code would have already transitioned the state to
+						// `CoverNoShows` before invoking this function, so that is not possible
+						// in practice.
+						//
+						// In the case of not-enough assignments, we want assignments to broadcast based
+						// on timer, so we treat it as though there are no uncovered no-shows.
+						if no_shows == 0 {
+							RequiredTranches::Pending(tranche_now)
+						} else if assignments < needed_approvals {
+							// This branch is only hit if there are no-shows before there
+							// are enough initial assignments.
+							RequiredTranches::Pending(tranche_now)
+						} else {
+							// This branch is never hit, as explained above.
+							RequiredTranches::Pending(tranche + no_shows as DelayTranche)
+						}
+					},
+				State::CoverNoShows(total_assignments, covered, covering, uncovered) =>
+					if covering == 0 && uncovered == 0 {
+						RequiredTranches::Exact(tranche, covered)
+					} else if total_assignments + covering + uncovered >= n_validators  {
```

Comment:
We considered this being less than `n_validators`, and could discover that being useful, but right now `n_validators` makes sense here.